File: libcli/CLI/Signature/FieldSig.cs
using System;

namespace CooS.Formats.CLI.Signature {

	public class FieldSig : MemberSig {

		public const byte FIELD = 0x06;

		public readonly CustomMod[] CustomMods;
		public readonly TypeSig Type;

		public static readonly SignatureFactory Factory = new FactoryImpl();

		internal static bool Predict(byte prologue) {
			return prologue==FIELD;
		}

		internal FieldSig(SignatureReader reader) : base(reader) {
			reader.ConfirmMark((ElementType)FIELD);
			CustomMods = SignatureUtility.ReadCustomMods(reader);
			Type = new TypeSig(reader);
		}

	}
}

File: kernel/signature/FieldSig.cpp
#include "FieldSig.h"


namespace Signature {

	FieldSig::FieldSig() {
	}

	FieldSig::~FieldSig() {
	}
	
	FieldSig::FieldSig(const FieldSig& value) : CustomMods(value.CustomMods), Type(value.Type) {
	}

	bool FieldSig::Parse(SignatureStream& _ss) {
		SignatureStream ss(_ss);
		if(ss.readByte()!=0x6/*FIELD*/) {
			return false;
		}
		while(true) {
			CustomMod cm;
			if(!cm.Parse(ss)) {
				break;
			}
			CustomMods.push_back(cm);
		}
		if(!Type.Parse(ss)) {
			return false;
		} else {
			ss.commit();
			return true;
		}
	}

}
